calicoctl/commands/ipam/check.go

```diff
@@ -0,0 +1,576 @@
+
+				// Found indicates whether the IP falls within an active IP pool.
+				found := false
+				parsedIP := net.ParseIP(ip)
+				for _, cidr := range activeIPPools {
+					if cidr.Contains(parsedIP) {
+						found = true
+						break
+					}
+				}
+				if !found {
+					if c.showProblemIPs {
+						for _, owner := range owners {
+							fmt.Printf("  %s in use by %v is not in any active IP pool.\n", ip, owner.FriendlyName)
+						}
+					}
+					nonCalicoIPs = append(nonCalicoIPs, ip)
+					continue
+				}
+				if c.showProblemIPs {
+					for _, owner := range owners {
+						fmt.Printf("  %s in use by %v and in active IPAM pool but has no IPAM allocation.\n", ip, owner.FriendlyName)
+					}
+				}
+				inUseButNotAllocatedIPs = append(inUseButNotAllocatedIPs, ip)
+			}
+		}
+		numProblems += len(nonCalicoIPs)
+		numProblems += len(inUseButNotAllocatedIPs)
+		fmt.Printf("Found %d in-use IPs that are not in active IP pools.\n", len(nonCalicoIPs))
+		fmt.Printf("Found %d in-use IPs that are in active IP pools but have no corresponding IPAM allocation.\n",
+			len(inUseButNotAllocatedIPs))
+		fmt.Println()
+	}
+
+	fmt.Printf("Check complete; found %d problems.\n", numProblems)
+
+	// Print out a machine readable report.
+	c.printReport()
+
+	return nil
+}
+
+func getWEPIPs(w apiv3.WorkloadEndpoint) ([]string, error) {
+	var ips []string
+	for _, a := range w.Spec.IPNetworks {
+		ip, err := normaliseIP(a)
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse IP (%s) of workload %s/%s: %w",
+				a, w.Namespace, w.Name, err)
+		}
+		ips = append(ips, ip)
+	}
+	return ips, nil
+}
+
+type Report struct {
+	// Version of the code that produced the report.
+	Version string `json:"version"`
+
+	// Important metadata.
+	ClusterGUID         string `json:"clusterGUID"`
+	DatastoreLocked     bool   `json:"datastoreLocked"`
+	ClusterInfoRevision string `json:"clusterInformationRevision"`
+	ClusterType         string `json:"clusterType"`
+
+	// Allocations is a map of IP address to list of allocation data.
+	Allocations map[string][]*Allocation `json:"allocations"`
+}
+
+func (c *IPAMChecker) printReport() {
+	r := Report{
+		Version:             c.version,
+		ClusterGUID:         c.clusterGUID,
+		ClusterType:         c.clusterType,
+		ClusterInfoRevision: c.clusterInfoRevision,
+		DatastoreLocked:     c.datastoreLocked,
+		Allocations:         c.allocations,
+	}
+	bytes, _ := json.MarshalIndent(r, "", "  ")
+	_ = ioutil.WriteFile(c.outFile, bytes, 0777)
+}
+
+// recordAllocation takes a block and ordinal within that block and updates
+// the IPAMChecker's internal state to track the allocation.
+func (c *IPAMChecker) recordAllocation(b *model.AllocationBlock, ord int) {
+	ip := b.OrdinalToIP(ord).String()
+	alloc := Allocation{IP: ip, Block: b, Ordinal: ord}
+
+	node := ""
+	blockAffinity := ""
+	if b.Affinity != nil {
+		affinity := *b.Affinity
+		if strings.HasPrefix(affinity, "host:") {
+			node = affinity[5:]
```

Comment:
If I'm understanding this correctly, this will include the `:` character in the node name. Is it intended for node to be `host:` here?